### ISSUE TITLE
Talent only abilities not updating and become unusable

### ISSUE THREAD

**Issue by wowregistering** [brittyazel/Neuron#211]
Prior to the most recent patch if you bound a key to a talent only ability, something like Death from Above or Marked for Death on a rogue, the ability change to a red '?' when untalented and change back to the proper icon when you retalent the ability.

After the recent update, if you follow the below you can replicate the issue easily.
1.  Select a Talent only ability in your Talent window
2. Drag the Ability to your bars, it will show the icon properly and function in combat
3. Untalent the ability, the icon will now stay the icon for the untalented ability, before the update it would turn into a red '?'
4. Retalent the ability, the icon will still have the icon of the ability that you just talented, untalented and retalented.

This causes confusion in the middle of combat, it makes you second guess what talents you have selected.  Every time you look at your bars you will see the icon for the ability you untalented being displayed as if you can currently use it.

This may be related to the other bug report I just made, the one about #showtooltip not changing the icon of buttons on the bar anymore.

**Comment by wowregistering** [comment]
Fixed with 0.9.38d

**Comment by brittyazel** [comment]
Hrm yeah, I'll have to think on that. Right now we basically just call a function that determines if the ability can be cast at the given time, and if it returns no webgray it out. But like you said that can be due to multiple reasons. Maybe I can split it up to make logical sense

**Comment by wowregistering** [comment]
Thinking about it more, I should probably not have reported this as a bug.  

When it 'fades out' the icon with the talent not selected it is actually behaving the same way as the default Blizzard interface, I was just used to Neuron bars turning it into a red '?' instead of greying it out.

**Comment by brittyazel** [comment]
Yeah the turning red is mostly for "out of range" indication. I'm going to mark this as an enhancement for now. I'd like to overhaul how we deal with talented abilities going forward though. I'd like to add some sort of indicator that the ability isn't talented. Perhaps a red overlay "/" or something of the sort.

**Comment by brittyazel** [comment]
I think I fixed this, can you confirm? For me, when I put Flourish on my bar and toggle the talent into it back and forth the ability goes grey as it should when I'm talented out of the ability

**Comment by wowregistering** [comment]
Also, it would seem you may have misinterpreted what I was talking about with the red buttons in my prior posts, I am talking about the button turning into a black background with a red question mark, not the red 'distance' color change.

http://1.bp.blogspot.com/-0bEZjGFAwp8/UFiOu9O2TTI/AAAAAAAAAE0/6IDrulAO_Aw/s1600/WoWScrnShot_091812_170517.jpg

The question mark icons in this random googled screenshot are the ones I was talking about.

**Comment by wowregistering** [comment]
Sorry for putting so many replies in separate posts, I had to roll back the version of my addon to 0.9.339g
https://wow.curseforge.com/projects/neuron/files/2649222

Using 1.0.0, 1.0.1, or 1.0.2c all gave me the error I posted above.

**Comment by brittyazel** [comment]
Yes the error that you listed above means that your profile cannot be migrated successfully to the 1.x spec. Some unfortunately had issues with the migration do to an unknown reason (usually old profiles).

In this scenario you unfortunately have 2 options, reset your profile for 1.x or stick to 0.9.x until you have a chance to reset the profile.

**Comment by wowregistering** [comment]
I set up my bars on my rogue and tried with both PvP and PvE Talent abilites, both seem to work properly now and fade out when untalented.

This also seems to have fixed the issue of untalented abilities turning into red question marks when first logging in or having not talented that ability since logging in.

A new issue I seem to be experiencing now with this update is the 'range' check seems to be very delayed now, sometimes up to 4 seconds to register that I am in range of something.

**Comment by wowregistering** [comment]
I have recorded a short clip displaying the range check issue I am experiencing.
Buttons
1: /cast Shadowstep
2: /cast [@mouseover] Shadowstep
3: /cast [@mouseover,exists] Shadowstep; Shadowstep
4: /cast Shadowstep/cast Kick
5: /cast Toxic Blade

Button 5, with Toxic Blade on it, is only being used so I have a button I can press that will not cause me to hit or teleport to the target while still displaying when I am entering a keypress that triggers an update of the range.

The white numbers below the bar are displaying my current distance to the target, when it displays 20-25 or less I am in range to Shadowstep and when it displays 25-30 or more then I am out of range to Shadowstep.

Video Timeline
0:00-0:02 I move my mouse over the target and there is about 1s of delay before it displays out of range, this is a mouseover only macro

0:04-0:07 I move in range of Shadowstep, wait a few seconds, and then press Toxic Blade to update the range display

0:07-0:11 I move out of range of Shadowstep, wait a few seconds, and then press Toxic Blade to update the range display

0:14-0:17 I move closer to the target until the range display updates to show Shadowstep as 'in range' while not pressing Toxic Blade to force it to update the range.

0:17-0:25 I move out of range of Shadowstep and wait for the range to display that I am no longer in range.  I am out of range at 22.3s and the range display updates at around 24.3s

0:25-0:33 I move in range of Shadowstep and wait for the range to display that I am in range without me pressing Toxic Blade to cause the range to be updated.  I am in range at 26s and it updates the display of my range at 33s.

https://gfycat.com/ColdSeriousBurro

I had an issue similar to this when using [harm] [help] conditionals on my Druid in the past but I just ignored it, never caused me any issues.  My Druids bars are not set up again yet, if you want me to test those on my Druid I can do that as soon as I get a chance.

**Comment by wowregistering** [comment]
Just started on my Druid and already have noticed issues with PvP talents.

If I spec into Cyclone and drag it to my bar it just displays a red question mark as the icon, the same for Thorns.

After messing around with this issue more, it appears to not be an issue with Druid, it seems like you have to have the ability available for use at least once before it will display the icon properly.

Steps Followed to reproduce
1: Fresh Profile with Warmode off
2: Talent into Cyclone and drag to bar, it will display as a question mark
After this step even if you log in and out it will not make the icon appear.

3: Turn War mode on, the icon will now appear.
After step 3 the icon will continue to display properly even if you untalent the ability or turn war mode off, continues functioning after relogs.

**Comment by brittyazel** [comment]
Yeah that's unavoidable sadly. I can't query the icon until the ability is physically available at least once. That's just a game limitation  The good news is that once the ability is cached, it's cached in memory indefinitely. So it's just that very first time
